clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -2952,6 +2952,163 @@ check_exit_found(void)
     return true;
 }
 
+bool
+check_kernel_trace_and_signal_markers(bool for_syscall)
+{
+#ifdef UNIX
+    trace_marker_type_t start_marker;
+    trace_marker_type_t end_marker;
+    uintptr_t file_type = OFFLINE_FILE_TYPE_SYSCALL_NUMBERS;
+    std::string test_type;
+    if (for_syscall) {
+        start_marker = TRACE_MARKER_TYPE_SYSCALL_TRACE_START;
+        end_marker = TRACE_MARKER_TYPE_SYSCALL_TRACE_END;
+        file_type |= OFFLINE_FILE_TYPE_KERNEL_SYSCALLS;
+        test_type = "Syscall";
+    } else {
+        start_marker = TRACE_MARKER_TYPE_CONTEXT_SWITCH_START;
+        end_marker = TRACE_MARKER_TYPE_CONTEXT_SWITCH_END;
+        test_type = "Context switch";
+    }
+    std::cerr << "Testing kernel trace for " << test_type << "\n";
+    // Matching kernel_event and kernel_xfer in the kernel trace.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_marker(TID_A, TRACE_MARKER_TYPE_FILETYPE, file_type),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_CACHE_LINE_SIZE, 64),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_PAGE_SIZE, 4096),
+            gen_instr(TID_A, /*pc=*/1),
+            // The syscall marker is needed for the syscall test but doesn't
+            // make any difference for the context switch test. We keep it
+            // for both to simplify test setup.
+            gen_marker(TID_A, TRACE_MARKER_TYPE_SYSCALL, 1),
+            gen_marker(TID_A, start_marker, 1),
+            gen_instr(TID_A, /*pc=*/10),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_KERNEL_EVENT, 11),
+            gen_instr(TID_A, /*pc=*/101),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_KERNEL_XFER, 102),
+            gen_instr(TID_A, /*pc=*/11),
+            gen_marker(TID_A, end_marker, 1),
+            gen_instr(TID_A, /*pc=*/2),
+            gen_exit(TID_A),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Extra signal kernel_event in the kernel trace.
```

Comment:
But it's not really a signal: an interrupt, right?

Reply:
This one is meant to represent an interrupt, yes. Reworded.

Reply:
This one is meant to represent an interrupt, yes. Reworded.